execution/unary/unary.go

```diff
@@ -40,26 +40,28 @@ func NewUnaryNegation(
 }
 
 func (u *unaryNegation) Series(ctx context.Context) ([]labels.Labels, error) {
-	var err error
-	u.once.Do(func() { err = u.loadSeries(ctx) })
-	if err != nil {
+	if err := u.loadSeries(ctx); err != nil {
 		return nil, err
 	}
 	return u.series, nil
 }
 
 func (u *unaryNegation) loadSeries(ctx context.Context) error {
-	vectorSeries, err := u.next.Series(ctx)
-	if err != nil {
-		return err
-	}
-	u.series = make([]labels.Labels, len(vectorSeries))
-	for i := range vectorSeries {
-		lbls := labels.NewBuilder(vectorSeries[i]).Del(labels.MetricName).Labels(nil)
-		u.series[i] = lbls
-	}
+	var err error
+	u.once.Do(func() {
+		var vectorSeries []labels.Labels
+		vectorSeries, err = u.next.Series(ctx)
+		if err != nil {
+			return
+		}
+		u.series = make([]labels.Labels, len(vectorSeries))
+		for i := range vectorSeries {
+			lbls := labels.NewBuilder(vectorSeries[i]).Del(labels.MetricName).Labels(nil)
+			u.series[i] = lbls
+		}
+		u.workers.Start(ctx)
+	})
 
-	u.workers.Start(ctx)
```

Comment:
I am wondering if we can remove workers completely, the overhead of using goroutines for this might not be worth the cost. 

Could you try running the benchmark for unary negation without workers (just inline the negation) to see what the difference will be?

Reply:
With workers:
```
$ go test . -run ^$ -bench BenchmarkOldEngineInstant/unary_negation
goos: linux
goarch: amd64
pkg: github.com/thanos-community/promql-engine/engine
cpu: 11th Gen Intel(R) Core(TM) i5-1135G7 @ 2.40GHz
BenchmarkOldEngineInstant/unary_negation/current_engine-8         	      34	 34777802 ns/op	 4071660 B/op	   56018 allocs/op
BenchmarkOldEngineInstant/unary_negation/new_engine-8             	      44	 24974904 ns/op	15283028 B/op	   65461 allocs/op
PASS
ok  	github.com/thanos-community/promql-engine/engine	15.752s
```

without workers:
```
$ go test . -run ^$ -bench BenchmarkOldEngineInstant/unary_negation
goos: linux
goarch: amd64
pkg: github.com/thanos-community/promql-engine/engine
cpu: 11th Gen Intel(R) Core(TM) i5-1135G7 @ 2.40GHz
BenchmarkOldEngineInstant/unary_negation/current_engine-8         	      34	 33501720 ns/op	 4068889 B/op	   55952 allocs/op
BenchmarkOldEngineInstant/unary_negation/new_engine-8             	      45	 23733152 ns/op	15274257 B/op	   65316 allocs/op
PASS
ok  	github.com/thanos-community/promql-engine/engine	16.221s
```

As the workers seem to not make a big difference, ill take them out!